### PR TITLE
Moderator: Mute or stop video for a peer globally

### PR DESCRIPTION
**Pull request by L4ky**
Hi, 
this PR enables MODERATORS to remotelly mute audio or stop video for a specific peer from Participant List.
This has a global effect, the user will be muted for everyone.

Please note that i intentionally only allowed muting and not unmuting, for privacy/security reasons.
Moderators will be able only to mute and stop video, but not resume them. The user has to reenable them.

This is my first time with React Framework so please be patient with me and my code. It can be improved for sure.

### CODE REVIEW

**Comment by tapionx** [comment]
yuhu! :partying_face:

File: server/lib/Room.js [hunk 1]
```diff
@@ -1378,8 +1378,16 @@ class Room extends EventEmitter
...
+				const { peerId } = request.data;
+
+				const mutePeer = this._peers[peerId];
+
+				if (!mutePeer)
+					throw new Error(`peer with id "${peerId}" not found`);
+
+				this._notification(mutePeer.socket, 'moderator:mute');
+
+				mutePeer.close();
```
Comment (havfo) [inline server/lib/Room.js]: It seems a bit rude to throw the peer out of the meeting ;)

Comment (havfo) [reply server/lib/Room.js]: Seems you beat me to it!

Comment (L4ky) [reply server/lib/Room.js]: Well if he's very noisy, he deserves it! :P

Already fixed, i'll test right now and let you know.

File: app/src/RoomClient.js [hunk 2]
```diff
@@ -1385,6 +1385,46 @@ export default class RoomClient
...
+		{
+			await this.sendRequest('moderator:mute', { peerId });
+		}
+		catch (error)
+		{
+			logger.error('mutePeer() failed: %o', error);
+		}
+
+		store.dispatch(
+			peerActions.setMutePeerInProgress(false));
```
Comment (havfo) [inline app/src/RoomClient.js:1405]: Missing `peerId`

File: app/src/RoomClient.js [hunk 3]
```diff
@@ -1385,6 +1385,46 @@ export default class RoomClient
...
+		{
+			await this.sendRequest('moderator:stopVideo', { peerId });
+		}
+		catch (error)
+		{
+			logger.error('stopPeerVideo() failed: %o', error);
+		}
+
+		store.dispatch(
+			peerActions.setStopPeerVideoInProgress(false));
```
Comment (havfo) [inline app/src/RoomClient.js:1425]: Missing `peerId`

**Comment by L4ky** [comment]
Ok fixed everything and tested with 3 clients: everything looks fine.

I apologize for all my mistakes. As i said this is the first time on React and i didn't dive in the code to see how it works. Now i have a much more clear idea about this part.

One question: when i disable a user camera the icon disappear. That means that webcamProducer is not anymore there.
Why this does not happen with micProducer?

**Comment by havfo** [comment]
I have tested here also, and it seems to work nicely now.

The reason is that we don't pause the video, we disable it. The audio is simply paused, not disabled. The `webcamConsumer` is then `null`.

**Comment by L4ky** [comment]
> I have tested here also, and it seems to work nicely now.
> 
> The reason is that we don't pause the video, we disable it. The audio is simply paused, not disabled. The `webcamConsumer` is then `null`.

Ok that confirmed what i was thinking.
Thank you so much!

**Comment by havfo** [comment]
The functionality now works, now I want some input on the UI part of it. There are now very many (too many) icons in the participant list. Also, green is probably not the right color.

**Comment by L4ky** [comment]
I agree with you. 
I think that we should choose a color for moderator tools and insert a label to indicate that those are moderator tools.
Plus i'd like to make the tooltip text more clear ( difference between local and global mute ).

What do you think about it?

**Comment by havfo** [comment]
Yes, that matches pretty much with my thoughts. I am just not sure how to make this difference clear, and a label takes up quite a bit of space.

We also might want to consider having all the moderator buttons in a container with a slightly different color, or border. Also we could have them in a dropdown on every peer with proper labels on every button (like the "More actions" button in the AppBar).

**Comment by so010** [comment]
> We also might want to consider having all the moderator buttons in a container with a slightly different color, or border. Also we could have them in a dropdown on every peer with proper labels on every button (like the "More actions" button in the AppBar).

What about https://material-ui.com/components/material-icons/  "build"  icon as a dropdown for moderator functions on each participant. Color could be blue or yellow.

**Comment by havfo** [comment]
Let's merge this as is, and I will modify it to be in a dropdown. Great work @L4ky !

**Comment by Alex-YddeR** [comment]
Global Unmute is also required for moderator as in many sessions moderator needs to have complete control over the users, other wise it will be time consuming process of making users to unmute themselves and moderator cannot tell every-time to unmute. i have been experienced with these kind of issues with many customers. 

Atleast after global mute we should have global unmute of individual participant at moderator.

**Comment by bpcurse** [comment]
@Alex-YddeR Allowing moderators to unmute individual or all users is a feature that has to be well documented and needs to be clearly Opt-In including a legal hint, or you might easily break the law operating such a server (regarding European GDPR)

**Comment by Alex-YddeR** [comment]
Yes i agree with you, if user muted himself both audio/video then there should be a request sent to the user to enable or not and it is up to user. 
Now when moderator mutes the user mic,  moderator can enable back by a click or a request can also be sent

but when Moderator mutes the user video and when he enables back, a mandatory request needs to sent to the user that host wants to enable your video, now user can deny or allow the request and its upto the user.

**Pull request merged**